music5000.js

```diff
@@ -0,0 +1,195 @@
+"use strict";
+
+// Code ported from Beebem (C to .js) by Jason Robson
+const AUDIO_BUFFER_SIZE = 256;
+
+const RAM_SIZE = 2048;
+const WAVE_TABLE_SIZE = 128;
+const WAVE_TABLES = 14;
+const NUM_CHANNELS = 16;
+const CHANNEL_REG_OFFSET = WAVE_TABLES * WAVE_TABLE_SIZE;
+const CHANNEL_ROW_SIZE = 128;
+
+// Control register bits
+const CTRL_STEREO_POS = (0xf);
+const CTRL_INVERT_WAVE = (1 << 4);
+const CTRL_MODULATE_ADJ = (1 << 5);
+
+// Data bits
+const DATA_SIGN = (1 << 7);
+const DATA_VALUE = (0x7f);
+const FREQ_DISABLE = (1 << 0);
+const NOT_FREQ_DISABLE = 0xfe;
+
+// Channel register sets
+const REG_SET_NORMAL = 0;
+const REG_SET_ALT = 1;
+
+
+export function Music5000(onBuffer) {
+	this._onBufferMusic5000 = onBuffer;
+
+	this.waveRam = new Uint8Array(RAM_SIZE);
+	this.phaseRam = new Uint32Array(NUM_CHANNELS);
+	this.cycleCount = 0;
+	this.curCh = 0;
+	this.activeRegSet = REG_SET_NORMAL;
+	this.sampleLeft = 0;
+	this.sampleRight = 0;
+	
+	this.chordBase = new Array(0, 8.25, 24.75, 57.75, 123.75, 255.75, 519.75, 1047.75 );
+	this.stepInc = new Array(0.5, 1.0, 2.0, 4.0, 8.0, 16.0, 32.0, 64.0 );
+
+	this.stereoLeft = new Array(0, 0, 0, 0, 0, 0, 0, 0, 100, 100, 100, 83, 67, 50, 33, 17);
+	this.stereoRight = new Array(100, 100, 100, 100, 100, 100, 100, 100, 0, 0, 0, 17, 33, 50, 67, 83);
+
+	this.D2ATable = new Uint16Array(128);
+	
+	this.sampleBuffer = new Float64Array(AUDIO_BUFFER_SIZE);
+	this.position = 0;
+
+	this.reset = function(hard)
+	{
+		if(hard)
+		{
+			console.log('Music 5000: initialisation');
+
+			// Build the D2A table
+			var i = 0;
+			for (var chord = 0; chord < 8; chord++)
+			{
+				var val = this.chordBase[chord];
+				for (var step = 0; step < 16; step++)
+				{
+					this.D2ATable[i] = parseInt(val * 4); // Multiply up to get an integer
```

Comment:
When I think of `parseInt` I'm usually thinking of parsing a string - if you need to get an integer value here you can do either: `Math.ceil()` or `Math.floor` or `val | 0` (which has the side-effect of forcing `val` to be an integer).

Reply:
Changed to Math.floor ;)